bridge-ffi/example/goffi/ffi.go

```diff
@@ -0,0 +1,169 @@
+		GetMessage() string
+	}
+}
+
+func (e ErrCore) Error() string { return e.Proto.GetMessage() }
+
+func (c *Core) RegisterWorker(in *bridgepb.RegisterWorkerRequest) error {
+	var resp bridgepb.RegisterWorkerResponse
+	req, inPtr, inLen, reqHandle := c.newRequest(in, &resp)
+	C.tmprl_register_worker(c.core, inPtr, inLen, reqHandle, callback)
+	<-req.ch
+	if err := resp.GetError(); err != nil {
+		return ErrCore{err}
+	}
+	return nil
+}
+
+func (c *Core) PollWorkflowActivation(
+	in *bridgepb.PollWorkflowActivationRequest,
+) (*workflowactivationpb.WorkflowActivation, error) {
+	var resp bridgepb.PollWorkflowActivationResponse
+	req, inPtr, inLen, reqHandle := c.newRequest(in, &resp)
+	C.tmprl_poll_workflow_activation(c.core, inPtr, inLen, reqHandle, callback)
+	<-req.ch
+	if err := resp.GetError(); err != nil {
+		return nil, ErrCore{err}
+	}
+	return resp.GetActivation(), nil
+}
+
+func (c *Core) newRequest(
+	reqProto interface{ Marshal() ([]byte, error) },
+	respProto proto.Message,
+) (req *request, inputPtr *C.uint8_t, inputLen C.size_t, reqHandle unsafe.Pointer) {
+	req = &request{core: c, resp: respProto, ch: make(chan *request, 1)}
+	if reqProto != nil {
+		var err error
+		if req.inputBytes, err = reqProto.Marshal(); err != nil {
+			panic(err)
+		}
+		inputPtr, inputLen = bytesPtrAndLen(req.inputBytes)
+	}
+	reqHandle = unsafe.Pointer(cgo.NewHandle(req))
+	return
+}
+
+func bytesPtrAndLen(b []byte) (*C.uint8_t, C.size_t) {
+	return (*C.uint8_t)(C.CBytes(b)), C.size_t(len(b))
+}
+
+func protoFromBytes(bytes *C.tmprl_bytes_t, p proto.Message) error {
+	// TODO(cretz): Why is the pointer not valid on an empty byte set?
+	if bytes.len == 0 {
+		return nil
+	}
+	in := (*[1<<30 - 1]byte)(unsafe.Pointer(bytes.bytes))
+	return proto.Unmarshal(in[:bytes.len], p)
+}
+
+type request struct {
+	// Expected to be passed in except for core init requests where it is set
+	core *Core
+	// This is set to nil on unmarshal error
+	resp proto.Message
+	// Sent on complete, should have buffer of 1 since non-blocking send is used
+	ch chan *request
+	// Only used to maintain a reference so it's not GC'd
+	inputBytes []byte
+}
+
+//export go_callback_core_init
+func go_callback_core_init(user_data C.uintptr_t, core *C.tmprl_core_t, bytes *C.tmprl_bytes_t) {
+	// Set core (which may be nil) then invoke normal callback
+	cgo.Handle(user_data).Value().(*request).core = &Core{core}
+	go_callback_core(user_data, bytes)
+}
+
+var initCallback = C.tmprl_core_init_callback(C.callback_core_init)
+var callback = C.tmprl_callback(C.callback_core)
+
+//export go_callback_core
+func go_callback_core(user_data C.uintptr_t, bytes *C.tmprl_bytes_t) {
+	h := cgo.Handle(user_data)
+	defer h.Delete()
+	req := h.Value().(*request)
+
+	// Deserialize
+	defer C.tmprl_bytes_free(req.core.core, bytes)
+	if req.resp != nil {
+		if err := protoFromBytes(bytes, req.resp); err != nil {
+			// TODO(cretz): Warn?
+			req.resp = nil
+		}
+	}
+
+	// Non-blocking channel send
+	select {
```

Comment:
Is this required with channel buffer size of 1?

Reply:
Not technically required, but it's safer to do a non-blocking send when you know you should never block a send